### ISSUE TITLE
KEDA should use TLS12 or above as minimum TLS version

### ISSUE THREAD

**Issue by JorTurFer** [kedacore/keda#4193]
### Report

Current, KEDA allows all TLS versions

### Expected Behavior

Not allowed broken TLS versions

### Actual Behavior

allowed broken TLS versions

**Comment by rmrfroot** [comment]
And limit TLS 1.2 cipher suites plz:
```
TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305

**Comment by JorTurFer** [comment]
> And limit TLS 1.2 cipher suites plz:
> 
> ```
> TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
> TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
> TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
> TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
> TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
> TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305
> ```

Why is this required? (I have no idea)

**Comment by rmrfroot** [comment]
> Why is this required? Is anyone broken? (I have no idea)

64-bit block cipher 3DES vulnerable to SWEET32 attack
To support only secure cipher suites.
https://ssl-config.mozilla.org/#server=traefik&version=2.1.2&config=intermediate&guideline=5.6

**Comment by JorTurFer** [comment]
Can this be solved using TLS 1.3 as default?

**Comment by rmrfroot** [comment]
If use only TLS 1.3 - of course. But all clients should also support it.

**Comment by JorTurFer** [comment]
> If use only TLS 1.3 - of course. But all clients should also support it.

I guess we can do it as we don't receive request do we don't have clients. The only requests we receive are from k8s api-server (which AFAIK supports it) and our own requests between components (which [already use TLS1.3](https://github.com/kedacore/keda/blob/main/pkg/metricsservice/utils/tls.go#L54)).
We are the clients of the upstreams, and I guess (better, I hope) support TLS1.3

Also taking this into account, maybe TLS1.2 could also be enough as we will use it only as clients, couldn't?
WDYT @zroubalik ?

**Comment by rmrfroot** [comment]
> The only requests we receive are from k8s api-server (which AFAIK supports it) and our own requests between components (which [already use TLS1.3](https://github.com/kedacore/keda/blob/main/pkg/metricsservice/utils/tls.go#L54)).

Looks great!

**Comment by zroubalik** [comment]
I am not sure if this not gonna affect scalers (in case they are using TLS for the connection). Might be safer to default on 1.2 here. Though I'd love to see 1.3 as default here. I don't have much knowledge in this area so I wonder what is the standard in opensource projects are k8s these days? 1.2 or 1.3? @rmrfroot by chance, do you know any details?

**Comment by JorTurFer** [comment]
BTW @rmrfroot , are you willing to contribute with the cipher suite improvements? We can create another issue to track those changes once the default version is merged.

**Comment by rmrfroot** [comment]
@zroubalik @JorTurFer Found this issue after one of my prod infrastructures with keda installed didn't pass security scans as a preparation to PCI DSS audit. So had to remove it. I'd really want to contribute if I had more experience in Go and more free time. But I can try to help.

**Comment by JorTurFer** [comment]
> Found this issue after one of my prod infrastructures with keda installed didn't pass security scans as a preparation to PCI DSS audit. So had to remove it

With [this PR](https://github.com/kedacore/keda/pull/4192), TLS1.2 will be the default minTLSVersion, but you could set to 1.3 just adding the environment variable if you need it to pass you audit. Could this be enough for your use case?

**Comment by rmrfroot** [comment]
@JorTurFer yes, this is a nice solution. So that we can test before breaking anything. Thanks!

**Comment by davidlanouette** [comment]
> And limit TLS 1.2 cipher suites plz: 
> ...

By default, Go only supports a limited set of "safe" cipher suites for TLSv1.2

[Here](https://cs.opensource.google/go/go/+/refs/tags/go1.20:src/crypto/tls/cipher_suites.go;l=53) is the list of CipherSuites that are supported by default (if you don't specify which CipherSuites to support) for TLSv1.2)  It changes regularly as vulnerabilities are found.

I would suggest you **don't** specify a list of suites because then you are in the business of knowing which ones have been exploited, and maintaining that list.  Let Google do that work for you - they are more likely to "get it right", and keep it up to date.

Setting the min version of TLS should be all you need to do.

**Comment by JorTurFer** [comment]
Wow,
Thanks a ton for the awesome clarification 🙇
